### PR TITLE
CDF-22673: add support for data modeling triggers in data workflows

### PR DESCRIPTION
**Pull request by vincent-cognite**
## Description
Please describe the change you have made.

## Checklist:
- [x] Tests added/updated.
- [x] Documentation updated. Documentation is generated from docstrings - these must be updated according to your change.
  If a new method has been added it should be referenced in [cognite.rst](https://github.com/cognitedata/cognite-sdk-python/blob/master/docs/source/cognite.rst) in order to generate docs based on its docstring.
- [x] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/cognite-sdk-python/blob/master/CHANGELOG.md).
- [x] Version bumped. If triggering a new release is desired, bump the version number in [_version.py](https://github.com/cognitedata/cognite-sdk-python/blob/master/cognite/client/_version.py) and [pyproject.toml](https://github.com/cognitedata/cognite-sdk-python/blob/master/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: cognite/client/data_classes/workflows.py [hunk 1]
```diff
@@ -1230,6 +1232,36 @@ def _load_trigger(cls, data: dict) -> Self:
         raise NotImplementedError()
 
 
+class WorkflowTriggerDataModelingQuery(CogniteObject):
```
Comment (haakonvt) [inline cognite/client/data_classes/workflows.py:1235]: Can you subclass `Query` instead?

File: cognite/client/data_classes/workflows.py [hunk 2]
```diff
@@ -1230,6 +1232,36 @@ def _load_trigger(cls, data: dict) -> Self:
...
+    """This class represents a data modeling trigger query."""
+
+    with_: dict[str, ResultSetExpression]
+    select: dict[str, Select]
+
+    def __init__(
+        self,
+        with_: dict[str, ResultSetExpression] = {},
+        select: dict[str, Select] = {},
```
Comment (haakonvt) [inline cognite/client/data_classes/workflows.py:1244]: Mutable default arguments can not be used here

File: cognite/client/data_classes/workflows.py [hunk 3]
```diff
@@ -1249,6 +1281,53 @@ def _load_trigger(cls, data: dict) -> WorkflowScheduledTriggerRule:
         return cls(cron_expression=data.get("cronExpression"))
 
 
+class WorkflowDataModelingTriggerRule(WorkflowTriggerRule):
+    """
+    This class represents a data modeling trigger rule.
+
+    Args:
+        data_modeling_query (WorkflowTriggerDataModelingQuery | None): The data modeling query of the trigger.
```
Comment (VerstraeteBert) [inline cognite/client/data_classes/workflows.py:1289]: Why is this nullable?

File: cognite/client/data_classes/workflows.py (hunk 1 above)
Comment (vincent-cognite) [reply cognite/client/data_classes/workflows.py:1235]: Query has `parameters` and `cursors` that we do not want to expose.

File: cognite/client/data_classes/workflows.py (hunk 2 above)
Comment (vincent-cognite) [reply cognite/client/data_classes/workflows.py:1244]: removed

File: cognite/client/data_classes/workflows.py (hunk 3 above)
Comment (vincent-cognite) [reply cognite/client/data_classes/workflows.py]: a mistake. Updated.

File: cognite/client/data_classes/workflows.py [hunk 4]
```diff
@@ -1240,8 +1240,8 @@ class WorkflowTriggerDataModelingQuery(CogniteObject):
 
     def __init__(
         self,
-        with_: dict[str, ResultSetExpression] = {},
-        select: dict[str, Select] = {},
+        with_: dict[str, ResultSetExpression],
```
Comment (VerstraeteBert) [inline cognite/client/data_classes/workflows.py:1243]: Why with_, typo?

**Review by VerstraeteBert — APPROVED** [review]
LGTM, one note

File: cognite/client/data_classes/workflows.py (hunk 4 above)
Comment (vincent-cognite) [reply cognite/client/data_classes/workflows.py:1243]: `with`is a reserved python keyword.
therefore following the Query parameter naming in query.py

File: cognite/client/data_classes/workflows.py [hunk 5]
```diff
@@ -1230,6 +1232,36 @@ def _load_trigger(cls, data: dict) -> Self:
...
+
+    with_: dict[str, ResultSetExpression]
+    select: dict[str, Select]
+
+    def __init__(
+        self,
+        with_: dict[str, ResultSetExpression],
+        select: dict[str, Select],
+    ) -> None:
+        super().__init__()
```
Comment (haakonvt) [inline cognite/client/data_classes/workflows.py:1246]:
```suggestion
```

File: cognite/client/data_classes/workflows.py [hunk 6]
```diff
@@ -1240,7 +1272,7 @@ class WorkflowScheduledTriggerRule(WorkflowTriggerRule):
 
     _trigger_type = "schedule"
 
-    def __init__(self, cron_expression: str | None = None) -> None:
+    def __init__(self, cron_expression: str | None) -> None:
         super().__init__()
```
Comment (haakonvt) [inline cognite/client/data_classes/workflows.py:1276]: These empty super-init calls look weird and may be removed by someone not looking too closely. I would advise to remove them and instead change the parent class:

```py
class WorkflowTriggerRule(CogniteObject, ABC):
    """This is the base class for all workflow trigger rules."""

    _trigger_type: ClassVar[str]

    def __init__(self) -> None:
        self.trigger_type = self._trigger_type
```

to e.g. this:
```py
class WorkflowTriggerRule(CogniteObject, ABC):
    """This is the base class for all workflow trigger rules."""

    _trigger_type: ClassVar[str]

    @property
    def trigger_type(self) -> str:
        return self._trigger_type
```

File: cognite/client/data_classes/workflows.py (hunk 1 above)
Comment (haakonvt) [reply cognite/client/data_classes/workflows.py:1235]: If its okay with you, I'll push a commit with a suggestion here (that we'll make a `_QueryCore` with `with_` and `select`, that both query classes can inherit from.

File: cognite/client/data_classes/workflows.py [hunk 7]
```diff
@@ -1249,6 +1281,51 @@ def _load_trigger(cls, data: dict) -> WorkflowScheduledTriggerRule:
...
+    ) -> None:
+        super().__init__()
+        self.data_modeling_query = data_modeling_query
+        self.batch_size = batch_size
+        self.batch_timeout = batch_timeout
+
+    @classmethod
+    def _load_trigger(cls, data: dict) -> WorkflowDataModelingTriggerRule:
+        return cls(
+            data_modeling_query=WorkflowTriggerDataModelingQuery.load(typing.cast(dict, data.get("dataModelingQuery"))),
```
Comment (haakonvt) [inline cognite/client/data_classes/workflows.py:1310]: Either the `"dataModelingQuery"` key is optional, and `WorkflowTriggerDataModelingQuery` needs to accept `dict | None`, or it is not optional, and you should not use `.get` (either way, no `typing.cast` should be here)

```suggestion
            data_modeling_query=WorkflowTriggerDataModelingQuery.load(data["dataModelingQuery"])),
```

**Review by haakonvt — COMMENTED** [review]
Mostly LGTM! A few suggestions from me 😄

File: cognite/client/data_classes/workflows.py (hunk 1 above)
Comment (haakonvt) [reply cognite/client/data_classes/workflows.py:1235]: At second thought, why not something like this:

```py
class WorkflowTriggerDataModelingQuery(Query):
    def __init__(
        self,
        with_: dict[str, ResultSetExpression],
        select: dict[str, Select],
    ) -> None:
        # parameters and cursors are not supported:
        super().__init__(with_, select_, parameters=None, cursors=None)
```

File: cognite/client/data_classes/workflows.py (hunk 6 above)
Comment (vincent-cognite) [reply cognite/client/data_classes/workflows.py:1276]: It seems the suggestion with the @property breaks

tests/utils.py

```
signature = inspect.signature(resource_cls.__init__)
```

Is there an obvious fix? I feel I'm spending more time reverse engineering a unit test than doing something useful at the moment 🙈

File: cognite/client/data_classes/workflows.py (hunk 7 above)
Comment (vincent-cognite) [reply cognite/client/data_classes/workflows.py]: fixed

File: cognite/client/data_classes/workflows.py [hunk 8]
```diff
@@ -1214,8 +1214,15 @@ class WorkflowTriggerRule(CogniteObject, ABC):
 
     _trigger_type: ClassVar[str]
 
-    def __init__(self) -> None:
-        self.trigger_type = self._trigger_type
+    @final
```
Comment (vincent-cognite) [inline cognite/client/data_classes/workflows.py:1217]: I didn't know about that one!

**Pull request merged**